### PR TITLE
Fix `bytes.concat("")`

### PR DESCRIPTION
**Pull request by cameel**
Discovered while working on #11539/#11580. The cause and fix are very similar, this is just a different case that was not in the original report.

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ Compiler Features:
 
 
 Bugfixes:
+ * Code Generator: Fix crash when passing an empty string to ``bytes.concat()``.
```
Comment (cameel) [inline Changelog.md]: Unlike with `bytes.concat(0)` in #11580, this is not our typical ICE, but something derived from `std::exception`:
```
Unknown exception during compilation: array::at: __n (which is 4294967295) >= _Nm (which is 32)
```
Not sure if I should call it "internal compiler error" here or not.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -1083,7 +1083,8 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 					targetTypes.emplace_back(argument->annotation().type);
 				else if (
 					auto const* literalType = dynamic_cast<StringLiteralType const*>(argument->annotation().type);
-					literalType && literalType->value().size() <= 32
+					literalType &&
+					0 < literalType->value().size() && literalType->value().size() <= 32
```
Comment (axic) [inline libsolidity/codegen/ExpressionCompiler.cpp]: I think `>= 1` is more readable for me at least:
```suggestion
					literalType->value().size() >= 1 && literalType->value().size() <= 32
```

Or
```suggestion
					!literalType->value().empty() && literalType->value().size() <= 32
```

File: test/libsolidity/semanticTests/array/concat/bytes_concat_empty_strings.sol [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        return bytes.concat(
+            bytes.concat(b),
+            bytes.concat(b, b),
+            bytes.concat("", b),
+            bytes.concat(b, "")
+        );
+    }
+
+    function g() public returns (bytes memory) {
+        return bytes.concat("", "abc", "");
```
Comment (axic) [inline test/libsolidity/semanticTests/array/concat/bytes_concat_empty_strings.sol]: Why not include the `hex""` and `unicode""` cases in the semantic tests too?

Comment (cameel) [reply test/libsolidity/semanticTests/array/concat/bytes_concat_empty_strings.sol]: Sure. Looking at the implementation I think it's unlikely that they behave any differently which is why I only covered them in syntax tests but I don't mind having them here either.

Comment (cameel) [reply test/libsolidity/semanticTests/array/concat/bytes_concat_empty_strings.sol]: Done. Changed to `bytes.concat("", "abc", hex"", "abc", unicode"")`.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 2 above)
Comment (cameel) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Agreed on `>= 1` vs `> 0`.

As to the order, I changed it but I find the original one more readable when I'm checking that a variable is in a certain range. It's closer to `0 < x < N` which is more natural to me.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 4]
```diff
@@ -2473,7 +2473,7 @@ string YulUtilFunctions::bytesConcatFunction(vector<Type const*> const& _argumen
 			targetTypes.emplace_back(argumentType);
 		else if (
 			auto const* literalType = dynamic_cast<StringLiteralType const*>(argumentType);
-			literalType && literalType->value().size() <= 32
+			literalType && literalType->value().size() >= 1 && literalType->value().size() <= 32
```
Comment (mijovic) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
			literalType && !literalType->value().empty() && literalType->value().size() <= 32
```

**Comment by mijovic** [comment]
Looks good in general, only one small suggestion

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 4 above)
Comment (cameel) [reply libsolidity/codegen/YulUtilFunctions.cpp]: Done.

**Review by mijovic — DISMISSED** [review]
Looks good

**Review by hrkrshnn — DISMISSED** [review]

**Comment by chriseth** [comment]
Does this allocate a size-0 memory array or does it a) not allocate at all or b) use the special "fixed" zero-length memory array?

If it is a) or b) I'm fine with merging as is.

**Comment by mijovic** [comment]
> Does this allocate a size-0 memory array or does it a) not allocate at all or b) use the special "fixed" zero-length memory array?
> 
> If it is a) or b) I'm fine with merging as is.

As far as I understood it skips `""` argument, so it should be option a) not allocating at all

**Comment by chriseth** [comment]
Please verify by looking at the generated code, @cameel

**Comment by cameel** [comment]
OK, I'll check but right away I can say that, at least in the code that collects types of arguments, `""` goes into the same branch as `bytes` (rather than fixed bytes).

**Comment by cameel** [comment]
> Does this allocate a size-0 memory array or does it a) not allocate at all or b) use the special "fixed" zero-length memory array?

I checked the IR output and looks like it does not allocate memory at all for string literals. It only allocates memory for the resulting bytes variable. The string literals are just embedded directly in Yul code.

I tried these 4 cases:
- `bytes memory b = bytes.concat("", "", "");`
- `bytes memory b = abi.encodePacked("", "", "");`
- `bytes memory b = bytes.concat("abc", "abc", "abc");`
- `bytes memory b = abi.encodePacked("abc", "abc", "abc");`

In cases with empty strings you get a function called `store_literal_in_memory_<string hash>(memPtr)` with empty body. For `"abc"` you get a function that does `mstore(pos, "abc")`.

In all four cases there's just a single pair of `allocate_unbounded()` + `finalize_allocation()` calls and that's for the result.

**Comment by cameel** [comment]
Oh, and between `bytes.concat()` and `abi.encodePacked()` variants the code is nearly identical. It differs only by insignificant details like inlining and naming (`abi.encodePacked` interprets the literal as `string`, and `bytes.concat()` as `bytes`).

**Review by chriseth — APPROVED** [review]

**Pull request merged**